Add a reference for Semgrepignore v2 (and v1 at the same time) (#1861)
semgrep/semgrep-docs: docs/semgrepignore-v2-reference.md

---
slug: semgrepignore-v2-reference
append_help_link: true
description: |
  Reference of the semgrepignore file fitering mechanism for Semgrep
hide_title: true

> title: Semgrepignore reference (v2)

Reviewer: ```suggestion
title: Semgrepignore v2 reference
```
Citation:
https://developers.google.com/style/parentheses

---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Semgrepignore reference (v2)

Reviewer: ```suggestion
# Semgrepignore v2 reference
```
Citation: 
https://developers.google.com/style/parentheses

> This document covers the Semgrepignore target filtering system that is

Reviewer: Do we need to capitalize semgrepignore?

Reviewer: To clarify: `gitignore` as a feature, not just the filename, **isn't** capitalized, so I'd prefer we follow their convention and also keep `semgrepignore` small caps.

Author: > Do we need to capitalize semgrepignore?

I think it's more logical. Writing "`gitignore`" is confusing imo.

My idea here is to capitalize "Gitignore" and "Semgrepignore" when referring to the whole feature as an ordinary proper noun (it's a proper noun because it was made up and refers to a unique thing). This implies using an ordinary font, not a monospace font. On the other hand, any text that's entered into the computer or read from it is quoted verbatim using a monospace font to indicate this. I think it's a common assumption. This applies to command names, full commands, file paths, and any text copy-pasted from a console. File names such as `.gitignore` or `.semgrepignore` as written as such (monospace font). What doesn't make sense to me is that the [Gitignore help page](https://git-scm.com/docs/gitignore) writes it as "`gitignore`" (lowercase, monospace font) even though there's no such thing as a `gitignore` command or file.

Reviewer: @mjambon That works for me. Let's keep it as:

- `.semgrepignore` referring to the filename
- Semgrepignore referring to the feature

> This document covers the Semgrepignore target filtering system that is

Reviewer: ```suggestion
This document covers the Semgrepignore v2 target filtering system that is
```

currently available with the `--experimental` option of the `semgrep`

> command ("v2"). It differs slightly from the legacy implementation which is

Reviewer: ```suggestion
command ("v2"). It differs from the legacy implementation which is
```

> command ("v2"). It differs slightly from the legacy implementation which is

Reviewer: ```suggestion
command. It differs slightly from the legacy implementation which is
```

referred to as "v1".

## The target filtering process

A `semgrep scan` command takes one or more scanning roots as
arguments. The default scanning root is the current folder, `.`.
Scanning roots are folders, individual files, or named pipes that should be
expanded into a list of regular files to be analyzed. Symbolic links are
allowed as scanning roots.

> Expanding a folder consists in listing its contents recursively with

Reviewer: ```suggestion
Expanding a folder consists of listing its contents recursively with
```

the following exceptions:

* Symbolic links other than the original scanning roots are ignored.
* In Git projects, Git submodules are ignored.
* Paths excluded via Semgrepignore patterns are ignored. Semgrepignore
  patterns can be of different sources which are detailed in the
  upcoming section.

The list of files obtained by expanding the scanning roots are called

> "target files". To obtain target files, Semgrep follows a

Reviewer: ```suggestion
**target files**. To obtain target files, Semgrep follows a
```
In general, we prefer to use **bold** for terms rather than "", which are used for titles of docs / media, dialog, or citations. 

https://developers.google.com/style/quotation-marks

number of fixed rules and some configurable filters.

For each scanning root, Semgrep infers a project root (v2 only). The
project root determines the location of applicable `.semgrepignore`
files as well as `.gitignore` files in Git projects. In v1 where is no
notion of a project root, the `.semgrepignore` file is unique and
looked up in the current folder.

<!-- TODO: explain project detection.
     Go over options to disable listing files using `git ls-files`
     while possibly still consulting the `.gitignore` files -- when we
     have an option for it. Right now we have only `--no-git-ignore`
     which is confusing and too coarse. I'd like to deprecate it as
     soon as we have finer-grained replacements.
-->

> As an experimental debugging aid, we provide the the `--x-ls` option

Reviewer: ```suggestion
As an experimental debugging aid, Semgrep provides the the `--x-ls` option
```

to list the target files. `--x-ls-long` additionally prints excluded

> files and a brief justification. Beware that these two options are

Reviewer: I suggest using the `caution` callout here for the "Beware..." sentence. 

Reference: https://docusaurus.io/docs/next/markdown-features/admonitions

likely to be renamed or change their behavior in the

> future. Meanwhile, typical usage is:

Reviewer: ```suggestion
future. Meanwhile, is typical usage is:
```

---
slug: semgrepignore-v2-reference
append_help_link: true
description: |
  Reference of the semgrepignore file fitering mechanism for Semgrep
hide_title: true
title: Semgrepignore v2 reference
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Semgrepignore v2 reference

This document covers the Semgrepignore **v2** target filtering system that is
currently available with the `--experimental` option of the `semgrep`
command. It differs from the legacy **v1** implementation.
referred to as "v1".

> 

Reviewer: @mjambon Can we add a section here about the minimum prerequisites to use this feature (like what version of Semgrep). Something like:

```suggestion
## Prerequisites

This feature requires Semgrep version v 1.10X.X. You can also use the `semgrep:latest` Docker image.
```

Author: I don't think we should because it's still experimental.

This behavior will become the default for CE users soon (with an option to use the old behavior). We'll specify the minimum semgrep version at that time.

(Git, Mercurial, etc.) indicating a project root.
If no project root is found this way, it
defaults to the scanning root itself if it is a folder or to its containing
folder if it is a regular file.

<!-- TODO: explain project detection.
     Go over options to disable listing files using `git ls-files`
     while possibly still consulting the `.gitignore` files -- when we
     have an option for it. Right now we have only `--no-git-ignore`
     which is confusing and too coarse. I'd like to deprecate it as
     soon as we have finer-grained replacements.
-->

:::caution
As an experimental debugging aid, Semgrep provides the `--x-ls` option
to list the target files. `--x-ls-long` additionally prints excluded
files and a brief justification. Beware that these two options are
likely to be renamed or change their behavior in the
future. Meanwhile, its typical usage is:
```
semgrep --x-ls
```
or
```
semgrep --x-ls --experimental
```
:::

## Sources of Semgrepignore patterns

A Semgrepignore pattern is a glob pattern that is matched by Semgrep
against file paths to determine whether these paths should be allowed or
disallowed as target files.

> Supported sources of Semgrepignore patterns are:

Reviewer 2: From the user perspective, I find "sources" to be an odd word here. They feel like uses to me---places that accept Semgrepignore patterns

Author: I rephrased this particular sentence. The term "sources" still occurs before and after this.

* command-line `--exclude` and `--include` filters;
* the `.semgrepignore` file in the current folder (v1 only);
* all the `.semgrepignore` files in the project (v2 only);
* all the `.gitignore` files in the project in Git projects (v2 only);
* default Semgrepignore patterns.

> These sources of filters are grouped into levels. At the end of a level,

Reviewer 2: I think the term I'm used to here is "precedence"? As in, `--exclude` and `--include` filters take precedence over `.gitignore` and `.semgrepignore`

Author: "precedence levels" I think is the complete term. thx

Author: Actually this is misleading: "`--exclude` and `--include` filters take precedence over `.gitignore` and `.semgrepignore`". The presence of `--exclude` or `--include` filters doesn't mean `.gitignore` and `.semgrepignore` will be ignored.
